les/fetcher.go

```diff
@@ -68,46 +68,47 @@ type fetcherPeer struct {
 	latest *announceData // The latest announcement sent from the peer
 
 	// These following two fields can track the latest announces
-	// from the peer with limited size for caching.
+	// from the peer with limited size for caching. We hold the
+	// assumption that all enqueued announces are td-monotonic.
 	announces     map[common.Hash]*announce // Announcement map
-	announceQueue *prque.Prque              // Announcement queue
+	announcesList []common.Hash             // FIFO announces list
 }
 
 // addAnno enqueues an new trusted announcement. If the queued announces overflow,
 // evict from the oldest.
-func (fp *fetcherPeer) addAnno(announce *announce) {
-	// Short circuit if the announce already exists. In normal case it should
-	// never happen since only monotonic announce is accepted. But the adversary
+func (fp *fetcherPeer) addAnno(anno *announce) {
+	// Short circuit if the anno already exists. In normal case it should
+	// never happen since only monotonic anno is accepted. But the adversary
 	// may feed us fake announces with higher td but same hash. In this case,
-	// ignore the announce anyway.
-	hash, number := announce.data.Hash, announce.data.Number
+	// ignore the anno anyway.
+	hash := anno.data.Hash
 	if _, exist := fp.announces[hash]; exist {
 		return
 	}
-	fp.announces[hash] = announce
-	fp.announceQueue.Push(hash, -int64(number))
+	fp.announces[hash] = anno
+	fp.announcesList = append(fp.announcesList, hash)
```

Comment:
I think this is going to grow the underlying array indefinitely and the slice over the array is just going to keep shifting. You could either use a circular index over a fixed size slice (efficient but more complex and less readable) or sometimes (after every Nth append) you should copy the list into a new array.

Reply:
fixed